---
title: "Reshape and clean data: wecare feb2016"
author: "Alexey Larionov"
output: html_document
params:
  interim_data: "/scratch/medgen/users/alexey/wecare_stat_sep2016/interim_data"
  subset: "std.1"
---

started: 01Mar2016  
last updated: 09Sep2016

# Summary

Select data for the requiested subset  
Explore and clean cases (remove odd and duplicated cases etc)  
Explore and clean phenoitype data, compile a single table with required phenotype annotations  
Save selected reshaped data

# start_section

```{r start_section}

# Start time
Sys.time()

# Do NOT clean-up at this stage
#rm(list=ls()) # this would remove params object!
graphics.off()

# Read parameters
interim_data_folder <- params$interim_data
data_subset <- params$subset

# Required libraries
library(tidyr) # for separate
library(dplyr) # for piping, filter, select etc
library(stringr) # for str_replace_all
library(ggplot2) # for some plots

# For debugging
#interim_data_folder <- "/scratch/medgen/users/alexey/wecare_aug2016/interim_data"
#data_subset <- "priority_genes_strict"
#setwd("/scratch/medgen/scripts/wecare_skat_08.16/scripts")

# Check subset
if (! data_subset %in% c("strict", "std.1", "std.2", "relaxed", "priority_genes_relaxed", "priority_genes_strict", "all_variants")){
  print("Wrong data subset")
  stop
}

```

# load_data

```{r load_data}

# Remove params object to avoid interference with load() function
rm(params)

interim_data_file <- paste(interim_data_folder, "s03_filter_by_effect.RData", sep="/")
load(file=interim_data_file)
ls()

if(data_subset == "strict"){
  vv.df <- vv.str
  kgen.df <- kgen.str
  exac.df <- exac.str
  genotypes.mx <- gt.str
  colnames(genotypes.mx) <- sub(".GT", "", colnames(genotypes.mx))
  interim_data_file <- paste(interim_data_folder, "s04_clean_pheno_str.RData", sep="/")
}

if(data_subset == "std.1"){
  vv.df <- vv.std.1
  kgen.df <- kgen.std.1
  exac.df <- exac.std.1
  genotypes.mx <- gt.std.1
  colnames(genotypes.mx) <- sub(".GT", "", colnames(genotypes.mx))
  interim_data_file <- paste(interim_data_folder, "s04_clean_pheno_std1_feb2016.RData", sep="/")
}

if(data_subset == "std.2"){
  vv.df <- vv.std.2
  kgen.df <- kgen.std.2
  exac.df <- exac.std.2
  genotypes.mx <- gt.std.2
  colnames(genotypes.mx) <- sub(".GT", "", colnames(genotypes.mx))
  interim_data_file <- paste(interim_data_folder, "s04_clean_pheno_std2_feb2016.RData", sep="/")
}

if(data_subset == "relaxed"){
  vv.df <- vv.rel
  kgen.df <- kgen.rel
  exac.df <- exac.rel
  genotypes.mx <- gt.rel
  colnames(genotypes.mx) <- sub(".GT", "", colnames(genotypes.mx))
  interim_data_file <- paste(interim_data_folder, "s04_clean_pheno_rel_feb2016.RData", sep="/")
}

if(data_subset == "priority_genes_relaxed"){
  vv.df <- vv.pri.rel
  kgen.df <- kgen.pri.rel
  exac.df <- exac.pri.rel
  genotypes.mx <- gt.pri.rel
  colnames(genotypes.mx) <- sub(".GT", "", colnames(genotypes.mx))
  interim_data_file <- paste(interim_data_folder, "s04_clean_pheno_pri_rel_feb2016.RData", sep="/")
}

if(data_subset == "priority_genes_strict"){
  vv.df <- vv.pri.str
  kgen.df <- kgen.pri.str
  exac.df <- exac.pri.str
  genotypes.mx <- gt.pri.str
  colnames(genotypes.mx) <- sub(".GT", "", colnames(genotypes.mx))
  interim_data_file <- paste(interim_data_folder, "s04_clean_pheno_pri_str_feb2016.RData", sep="/")
}

if(data_subset == "all_variants"){
  genotypes.mx <- gt.mx
  colnames(genotypes.mx) <- sub(".GT", "", colnames(genotypes.mx))
  interim_data_file <- paste(interim_data_folder, "s04_clean_pheno_all_feb2016.RData", sep="/")
}

rm(gt.mx, gq.mx, dp.mx, ref.mx, alt.mx)

rm(vv.str, gt.str, gq.str, dp.str, ref.str, alt.str, kgen.str, exac.str)
rm(vv.std.1, gt.std.1, gq.std.1, dp.std.1, ref.std.1, alt.std.1, kgen.std.1, exac.std.1)
rm(vv.std.2, gt.std.2, gq.std.2, dp.std.2, ref.std.2, alt.std.2, kgen.std.2, exac.std.2)
rm(vv.rel, gt.rel, gq.rel, dp.rel, ref.rel, alt.rel, kgen.rel, exac.rel)

rm(vv.pri.rel, gt.pri.rel, gq.pri.rel, dp.pri.rel, ref.pri.rel, alt.pri.rel, kgen.pri.rel, exac.pri.rel)
rm(vv.pri.str, gt.pri.str, gq.pri.str, dp.pri.str, ref.pri.str, alt.pri.str, kgen.pri.str, exac.pri.str)

ls()

# I do NOT convert_dfs_to_tables because it loses row names. 
# However, it could be considered later, 
# when row names handling in tables is fixed. 

```

# check_data

```{r check_data}

dim(covar.df)
str(covar.df)
covar.df[1:5,1:5]

dim(demographics.df)
str(demographics.df)
demographics.df[1:5,1:5]

dim(samples.df)
str(samples.df)
samples.df[1:5,]

dim(vv.df)
str(vv.df)
vv.df[1:5,1:5]

dim(kgen.df)
str(kgen.df)
kgen.df[1:5,1:5]

dim(exac.df)
str(exac.df)
exac.df[1:5,1:5]

dim(genotypes.mx)
class(genotypes.mx)
genotypes.mx[10:25,1:5]

# Check consistence of rownames
sum(rownames(genotypes.mx) != rownames(vv.df), na.rm=TRUE)
sum(rownames(genotypes.mx) != rownames(kgen.df), na.rm=TRUE)
sum(rownames(genotypes.mx) != rownames(exac.df), na.rm=TRUE)

```

# reshape_variants_annotations

```{r reshape_variants_annotations}

# Move variants IDs to explicit variable
vv.df <- cbind(rownames(vv.df),vv.df)
"VarID" -> colnames(vv.df)[1]

# select relevant variants annotations
variants.df <- 
  vv.df %>% 
  select(VarID, SYMBOL, TYPE, CHROM, POS, REF, ALT, AC, AF, AN, 
         Consequence, SIFT, PolyPhen, CLIN_SIG, 
         GMAF, EUR_MAF, ALT_frequency_in_1k_95, bc_risk, bc_somatic, dna_repair, es_related)

rm(vv.df)

# Explore cases with confusing GMAF/EUR_MAF containing & symbol (as inserted by VEP)
nrow(variants.df %>% filter(grepl("&",GMAF)))
nrow(variants.df %>% filter(grepl("&",EUR_MAF)))
variants.df %>% 
  filter(grepl("&",EUR_MAF)) %>% 
  select(c(1:6), GMAF, EUR_MAF) %>% 
  top_n(10,EUR_MAF)

# There are even 3 variants with triplicated values separated by &
variants.df[c(1235, 11405, 15894),c("GMAF", "EUR_MAF")]

# Split cases with confusing EUR_MAF containing "&"" symbol
variants.df <- 
  variants.df %>% 
  separate(EUR_MAF, c("EUR_MAF1", "EUR_MAF2"), "&") 

# All duplicated values are the same
sum(variants.df$EUR_MAF1 != variants.df$EUR_MAF2, na.rm=TRUE)

# Remove one of the duplicates
variants.df <- 
  variants.df %>% 
  select(-EUR_MAF2)

# split GMAF and EUR_MAF1
variants.df <- 
  variants.df %>% 
  separate(GMAF, c("GMAF_Allele", "GMAF_Fraction"), ":") %>% 
  separate(EUR_MAF1, c("EUR_MAF_Allele", "EUR_MAF_Fraction"), ":") 

# Convert columns to numeric etc
as.numeric(variants.df$GMAF_Fraction) -> variants.df$GMAF_Fraction
as.numeric(variants.df$EUR_MAF_Fraction) -> variants.df$EUR_MAF_Fraction

# Change "" to NAs
NA -> variants.df$GMAF_Allele[variants.df$GMAF_Allele == ""]
NA -> variants.df$EUR_MAF_Allele[variants.df$EUR_MAF_Allele == ""]
NA -> variants.df$SIFT[variants.df$SIFT == ""]
NA -> variants.df$PolyPhen[variants.df$PolyPhen == ""]
NA -> variants.df$CLIN_SIG[variants.df$CLIN_SIG == ""]

# Convert chars to factors
as.factor(variants.df$GMAF_Allele) -> variants.df$GMAF_Allele
as.factor(variants.df$EUR_MAF_Allele) -> variants.df$EUR_MAF_Allele

# Change "true" to TRUE
sum(variants.df$ALT_frequency_in_1k_95 == "true", na.rm = TRUE)
variants.df <- 
  variants.df %>% 
  mutate(frequent_in_1k=as.logical(
    str_replace_all(ALT_frequency_in_1k_95, "true", TRUE))) %>% 
  select(-ALT_frequency_in_1k_95)
sum(variants.df$frequent_in_1k, na.rm=TRUE)

# Split SIFT
variants.df <- 
  variants.df %>% 
  mutate(SIFT_call=sub("\\(.*\\)","",SIFT)) %>% 
  mutate(SIFT_score=as.numeric(
    sub(".*\\(","", sub("\\)","",SIFT)))) %>% 
  select(-SIFT)

# Split PolyPhen
variants.df <- 
  variants.df %>% 
  mutate(PolyPhen_call=sub("\\(.*\\)","",PolyPhen)) %>% 
  mutate(PolyPhen_score=as.numeric(
    sub(".*\\(","", sub("\\)","",PolyPhen)))) %>% 
  select(-PolyPhen)

# --- Rearrange priority genes columns --- #

bc_risk_v <- rep("", nrow(variants.df))
"bc_risk" -> bc_risk_v[variants.df$bc_risk]

bc_somatic_v <- rep("", nrow(variants.df))
"bc_somatic" -> bc_somatic_v[variants.df$bc_somatic]

dna_repair_v <- rep("", nrow(variants.df))
"dna_repair" -> dna_repair_v[variants.df$dna_repair]

es_related_v <- rep("", nrow(variants.df))
"es_related" -> es_related_v[variants.df$es_related]

priority_genes <- paste(bc_risk_v,bc_somatic_v,dna_repair_v,es_related_v, sep=",")
priority_genes <- sub(",{2,}", ",", priority_genes)
priority_genes <- sub("^,", "", priority_genes)
priority_genes <- sub(",$", "", priority_genes)

variants.df <- cbind(variants.df,priority_genes)
variants.df <- variants.df %>% select(- bc_risk, - bc_somatic, - dna_repair, - es_related)
NA -> variants.df[variants.df$priority_genes == "", "priority_genes"]

rm(priority_genes, bc_risk_v, bc_somatic_v, dna_repair_v, es_related_v)

# Check variants table
dim(variants.df)
str(variants.df)
variants.df[1:5,1:5]

```

# reshape_covar

```{r reshape_covar}

dim(covar.df)
colnames(covar.df)

attach(covar.df)

# chemo is the same as chemo_self_mra, no missed values
sum(chemo != chemo_self_mra)

# One missed case in hormone/ hormone_self_mra
sum(is.na(hormone))
sum(is.na(hormone_self_mra))

# hormone has 10 differences from hormone_self_mra
sum(hormone != hormone_self_mra, na.rm=TRUE)

# status is the same as status2, no missed values
sum(status != status2)

# Race is always 0, no missed data
sum(race)

# Dismiss split 495 vs 3, no missed data
sum(dsmiss)

# good_location split 86 vs 412, no missed data
sum(good_location)

# Deleterious split 480 vs 18, no missed data
sum(Deleterious)

# Deleterious split 158 vs 340, no missed data
sum(family_history)

# XRTBrCHAR is the same as XRTBreast, no missed values
sum(XRTBrCHAR != XRTBreast)

# sub_dx_age_cat split 169 vs 329, no missed data
sum(sub_dx_age_cat)

# dose
summary(dose)

detach(covar.df)

# Keep only selected annotations
selected_annotations <- c(
  "labid","cc","sub_dx_age","offset",
  "chemo","hormone","XRTBreast","dose",
  "Eigen_1","Eigen_2","Eigen_3","Eigen_4","Eigen_5")

# Keep only selected annotations
covar.df <- covar.df[,selected_annotations]
rm(selected_annotations)

# change XRTBreast to xrt
"xrt" -> colnames(covar.df)[7]

# Check result
dim(covar.df)
str(covar.df)
covar.df[1:5,1:5]

```

# explore_demographics_table

## exclude_odd_cases

```{r exclude_odd_cases}

# Outlook
dim(demographics.df)
colnames(demographics.df)
demographics.df[1:10,1:5]

# List of odd cases
odd_cases=c("201558","201921","202698","21IAno","387460",
            "389192","58IR1+","60IA1+","92IRno","98IAno")

# Show odd cases
demographics.df %>% 
  filter(Subject_ID %in% odd_cases) %>% 
  select(1:5)

# Exclude odd cases
pheno.df <- demographics.df %>% filter(! Subject_ID %in% odd_cases)
dim(pheno.df)
pheno.df[1:10,1:5]

```

## exclude_duplicated_cases

```{r exclude_duplicated_cases}

duplicated_cases <- 
  pheno.df %>% 
  filter(duplicated(Subject_ID)) %>% 
  select(Subject_ID)

duplicated_cases <- as.vector(duplicated_cases[,1])
duplicated_cases

# Phenotype annotations for the duplicate record are identical
pheno.df %>% 
  filter(Subject_ID %in% duplicated_cases) %>% 
  select(1:5)

pheno.df <- pheno.df %>% filter(!duplicated(Subject_ID))
dim(pheno.df)
pheno.df[1:10,1:5]

attach(pheno.df)

```

## x_y_id_cc_fam

```{r x_y_id_cc_fam}

# Case-control status

ID.x[1:5]
sum(ID.x != ID.y)

cc.x[1:20]
sum(cc.x != cc.y)

ggplot(pheno.df) + 
  aes(as.factor(cc.x), fill=as.factor(cc.x)) +
  geom_bar(colour="black", alpha=0.3)

table(cc.x)

# Lables x vs y

labid.x[1:5]
sum(as.vector(labid.x) != as.vector(labid.y))

# Familial history

fam_hist[1:5]
sum(fam_hist != family_history.y)
sum(as.numeric(sub("none",0,sub("1\\+",1,family_history.x))) != family_history.y)

ggplot(pheno.df) + 
  aes(as.factor(family_history.x), fill=as.factor(cc.x)) +
  geom_bar(colour="black", alpha=0.3)

table(family_history.x)
table(family_history.y)

# No disbalance in familial history between cases and controls

ggplot(pheno.df) + 
  aes(as.factor(cc.x), fill=as.factor(family_history.x)) +
  geom_bar(colour="black", alpha=0.3)

table(cc.y, family_history.y)
fisher.test(table(cc.y, family_history.y))

```

## x_y_ages

```{r x_y_ages}

# 1st Ds age (?)

sub_dx_age.x[1:5]
sum(sub_dx_age.x != sub_dx_age.y)

ggplot(pheno.df) + 
  aes(sub_dx_age.x, fill=as.factor(cc.x)) +
  geom_histogram(colour="black", alpha=0.3, binwidth = 1)
# Consistent with 55 yr selection criteria

# No differences between cases and controls by age
t.test(sub_dx_age.x~cc.x)

ggplot(pheno.df) + 
  aes(sub_dx_age.x, fill=as.factor(cc.x)) +
  geom_density(colour="black", alpha=0.3)

# Compare to Refage (age at 2nd ds???)
ggplot(pheno.df, colour="black") + 
  geom_density(aes(refage), fill="red", alpha=0.3) +
  geom_density(aes(sub_dx_age.x), fill="blue", alpha=0.3)

# No differences between cases and controls by refage
t.test(refage~cc.x)

ggplot(pheno.df) + 
  aes(refage, fill=as.factor(cc.x)) +
  geom_density(colour="black", alpha=0.3)

# Categorical age codings

dxyr_stratum[1:5]
age_stratum1[1:5]
sub_dx_age_cat[1:5]

# Unknown offset
# consistent with 1+ year between the 1st and 2nd tumours (selection criteria)

offset.x[1:5]
sum(offset.x != offset.y)
hist(offset.x) 

# Compare to rstime (~age at ds - age at 2nd ds???)
ggplot(pheno.df, colour="black") + 
  geom_density(aes(offset.x), fill="red", alpha=0.3) +
  geom_density(aes(rstime), fill="blue", alpha=0.3)

min(offset.x)
min(rstime)

```

## x_y_any_treatment

```{r x_y_any_treatment}

# Any treatment 
treatment.x[1:5]
treatment.y[1:5]
sum(treatment.x != treatment.y)

ggplot(pheno.df) + 
  aes(as.factor(treatment.x), fill=as.factor(cc.x)) +
  geom_bar(colour="black", alpha=0.3)

# There was more treatment in controls
table(cc.x, treatment.x)
fisher.test(table(cc.x, treatment.x))

ggplot(pheno.df) + 
  aes(as.factor(cc.x), fill=as.factor(treatment.x)) +
  geom_bar(colour="black", alpha=0.3)

# for comparison (neither will be used later)
chemo_hormone[1:5] 

```

## x_y_hormonal_treatment

```{r x_y_hormonal_treatment}

# Some missed data and discrepansies in endocrine treatment
# (when compare hormone, hormone_self_mra.x and hormone_self_mra.y)
hormone[1:5]
hormone_self_mra.x[1:5]
hormone_self_mra.y[1:5]
sum(hormone_self_mra.x != hormone_self_mra.y, na.rm=TRUE)
sum(hormone != hormone_self_mra.x, na.rm=TRUE)
sum(is.na(hormone))
sum(is.na(hormone_self_mra.x))
sum(is.na(hormone_self_mra.y))

ggplot(pheno.df) + 
  aes(as.factor(hormone), fill=as.factor(cc.x)) +
  geom_bar(colour="black", alpha=0.3)

# There was no disbalance in cases and controls by hormonal treatment
table(cc.x,hormone)
fisher.test(table(cc.x, hormone))

pheno.df %>% 
  filter(!is.na(hormone)) %>% 
  ggplot() + 
    aes(as.factor(cc.x), fill=as.factor(hormone)) +
    geom_bar(colour="black", alpha=0.3)

# Hormone vs er: 
# 64% er+ves were not given hormones:
# old cohort, young patients, cytotoxic instead ...?
# 11 er-ves were hiven hormones??
table(er1, hormone)
sum(is.na(er1))

```

## x_y_cytotoxic_treatment

```{r x_y_cytotoxic_treatment}

# No missed data or discrepansies in cytotoxic treatment
# (when compare chemo, chemo_self_mra.x and chemo_self_mra.y)
chemo[1:5]
sum(chemo_self_mra.x != chemo_self_mra.y)
sum(chemo != chemo_self_mra.x)

# Unbalanced cases and controls in treated/non-treated by cytotoxics
ggplot(pheno.df) + 
  aes(as.factor(chemo), fill=as.factor(cc.x)) +
  geom_bar(colour="black", alpha=0.3)

# There was more chemo in controls
table(cc.x, chemo)
fisher.test(table(cc.x, chemo))

ggplot(pheno.df) + 
  aes(as.factor(cc.x), fill=as.factor(chemo)) +
  geom_bar(colour="black", alpha=0.3)

```

## x_y_x_ray_treatment

```{r x_y_x_ray_treatment}

# No missed data or discrepansies in x-ray treatment
# (when compare XRTBrCHAR, XRTBreast.x and XRTBreast.y)
XRTBrCHAR[1:5]
sum(XRTBreast.x != XRTBreast.y)
sum(XRTBrCHAR != XRTBreast.x)

# Unbalanced cases and controls in treated/non-treated by radiotherapy
ggplot(pheno.df) + 
  aes(as.factor(XRTBrCHAR), fill=as.factor(cc.x)) +
  geom_bar(colour="black", alpha=0.3)

# There was more x_ray_treatment in controls
table(cc.x, XRTBrCHAR)
fisher.test(table(cc.x, XRTBrCHAR))

ggplot(pheno.df) + 
  aes(as.factor(cc.x), fill=as.factor(XRTBrCHAR)) +
  geom_bar(colour="black", alpha=0.3)

```

## x_y_eigenvectors

Ethnisity: todo
PCI co-plotting with known populations
Recalculate eigenvectors?

```{r x_y_eigenvectors}

# Eigenvectors - duplicated
sum(Eigen_1.x != Eigen_1.y)
sum(Eigen_2.x != Eigen_2.y)
sum(Eigen_3.x != Eigen_3.y)
sum(Eigen_4.x != Eigen_4.y)
sum(Eigen_5.x != Eigen_5.y)

```

## x_y_mixed_duplicated_fields

```{r x_y_mixed_duplicated_fields}

# setno (unknown field)
setno.x[1:5]
sum(setno.x != setno.y)
t.test(setno.x~cc.x)

ggplot(pheno.df) +
  aes(setno.x, fill=as.factor(cc.x)) +
  geom_density(colour="black", alpha=0.3)

# Registry
registry.x[1:5]
sum(as.vector(registry.x) != as.vector(registry.y))
fisher.test(table(cc.x,registry.x))

ggplot(pheno.df) +
  aes(registry.x, fill=as.factor(cc.x)) +
  geom_bar(colour="black", alpha=0.3)

# Good location (unknown field)
good_location.x[1:5]
sum(good_location.x != good_location.y)
fisher.test(table(cc.x,good_location.x))

ggplot(pheno.df) +
  aes(good_location.x, fill=as.factor(cc.x)) +
  geom_bar(colour="black", alpha=0.3)

# Status is equal to cc
status.x[1:5]
sum(status.x != status.y)
table(cc.x, status.x)

sum(status2.x != status2.y)
sum(status.x != status2.y)

# Race
sum(race.x != race.y)
sum(sub_race != race.y)
sum(sub_race)

```

## x_ray_treatment

```{r x_ray_treatment}

# dose
dose[1:5]
class(dose)
summary(dose)

ggplot(pheno.df) + 
  aes(as.factor(dose), fill=as.factor(cc.x)) +
  geom_bar(colour="black", alpha=0.3)

# More x-ray treatment in controls
fisher.test(table(cc.x, dose))

# Balanced doses in x-ray treated cases and controls
pheno.df %>% 
  select(cc.x,dose) %>% 
  filter(dose %in% c("ge 1Gy","ls 1Gy")) %>% 
  table %>% 
  fisher.test

# No missed data in dose
sum(is.na(dose))

# dose_caseloc
dose_caseloc[1:5]
class(dose_caseloc)

ggplot(pheno.df) +
  aes(dose_caseloc, fill=as.factor(cc.x)) +
  geom_histogram(colour="black", alpha=0.3)

ggplot(pheno.df) +
  aes(dose_caseloc, fill=as.factor(cc.x)) +
  geom_density(colour="black", alpha=0.3)

sum(dose_caseloc == 0, na.rm=TRUE)
sum(dose_caseloc != 0 & dose_caseloc < 1, na.rm=TRUE)
sum(dose_caseloc >=1, na.rm=TRUE)
sum(is.na(dose_caseloc))

# Numerically coded dose (0,1,2)
dose_num[1:5]
class(dose_num)

ggplot(pheno.df) +
  aes(dose_num, fill=as.factor(cc.x)) +
  geom_bar(colour="black", alpha=0.3)

table(dose_num)
sum(is.na(dose_num))

# Disbalance btw cases and controls (higher doses in controls?)
table(cc.x,dose_num)
fisher.test(table(cc.x,dose_num))

ggplot(pheno.df) +
  aes(as.factor(cc.x), fill=as.factor(dose_num)) +
  geom_bar(colour="black", alpha=0.3)

# Avedose (unknown dose)
avedose[1:5]
class(avedose)

ggplot(pheno.df) +
  aes(avedose, fill=as.factor(cc.x)) +
  geom_histogram(colour="black", alpha=0.3)

ggplot(pheno.df) +
  aes(avedose, fill=as.factor(cc.x)) +
  geom_density(colour="black", alpha=0.3)

# Disbalance between cases and controls
t.test(avedose~cc.x)

# More balanced amongst treated?
pheno.df %>% filter(avedose!=0) %>% 
  t.test(data=.,avedose~cc.x)

```

## hormonal_treatment

```{r hormonal_treatment}

# Inconsistent data on endocrine treatment
# Apparently 9 patients did not received hormones, 
# while receiving tamoxifen??
table(hormone,tmx)

# horm_tmx is different from tmx
# Apparently 0 is no hormonal treatment
# What is coded as 1 and 2 ??
horm_tmx[1:5]

ggplot(pheno.df) +
  aes(as.factor(horm_tmx), fill=as.factor(cc.x)) +
  geom_bar(colour="black", alpha=0.3)

table(horm_tmx)
sum(is.na(horm_tmx))

# Tmx
tmx[1:5]
barplot(table(tmx))
table(tmx)
sum(is.na(tmx))

# Treatment by age: hormones more often are given to older patients
x <- !is.na(hormone)
ggplot(pheno.df[x,]) + 
  aes(x = sub_dx_age.x, fill = as.factor(hormone)) +
  geom_density(colour="black", alpha=0.3)
rm(x)

t.test(sub_dx_age.x~hormone)

```

## cytotoxic_treatment

```{r cytotoxic_treatment}

# Reminder: chemo is identical to chemo.x/y
chemo[1:5]
ggplot(pheno.df) +
  aes(as.factor(chemo), fill=as.factor(cc.x)) +
  geom_bar(colour="black", alpha=0.3)
table(chemo)
sum(is.na(chemo))

# Treatment by age: cytotoxic are more often given to younger patients
ggplot(pheno.df) + 
  aes(x = sub_dx_age.x, fill = as.factor(chemo)) +
  geom_density(colour="black", alpha=0.3)

t.test(sub_dx_age.x~chemo)

# cmf uses wrong encodings
cmf[1:5]
ggplot(pheno.df) +
  aes(as.factor(cmf), fill=as.factor(cc.x)) +
  geom_bar(colour="black", alpha=0.3)
table(cmf)
sum(is.na(cmf))

# cmf012 is numeric cmf
# 0= none, 1=cmf, 2=other (the order is not meaningful)
cmf_012[1:5]
ggplot(pheno.df) +
  aes(as.factor(cmf_012), fill=as.factor(cc.x)) +
  geom_bar(colour="black", alpha=0.3)
table(cmf_012)
sum(is.na(cmf_012))

# CMF is coded better than cmf
CMF[1:5]
ggplot(pheno.df) +
  aes(as.factor(CMF), fill=as.factor(cc.x)) +
  geom_bar(colour="black", alpha=0.3)
table(CMF)
sum(is.na(CMF))

# CMF vs age
ggplot(pheno.df) + 
  aes(x = sub_dx_age.x, fill = as.factor(CMF)) +
  geom_density(colour="black", alpha=0.3)

# CMF is different from cmf
sum(as.vector(cmf) != as.vector(CMF))

# CMF only
CMF_Only[1:5]
ggplot(pheno.df) +
  aes(as.factor(CMF_Only), fill=as.factor(cc.x)) +
  geom_bar(colour="black", alpha=0.3)
table(CMF_Only)
sum(is.na(CMF_Only))

# # Treatment by age: patients given cytotoxics are younger than those given hormones
ggplot() + 
  geom_density(data=pheno.df[chemo==1,], aes(sub_dx_age.x), colour="black", fill="red", alpha=0.25) +
  geom_density(data=pheno.df[hormone==1,], aes(sub_dx_age.x), colour="black", fill="blue", alpha=0.25)

t.test(sub_dx_age.x[chemo==1],sub_dx_age.x[hormone==1])

```

## pathology

```{r pathology}

# Stage
stage_fd[1:5]
summary(as.factor(stage_fd))
ggplot(pheno.df) + 
  aes(as.factor(stage_fd), fill=as.factor(cc.x)) + 
  geom_bar(colour="black", alpha=0.3)

# No significant differences by stage between cases and controls
# However, tendency to higher stage in controls, 
# which might had caused more cytotoxics and x-ray in controls
fisher.test(table(cc.x, stage_fd))
ggplot(pheno.df) + 
  aes(as.factor(cc.x), fill=as.factor(stage_fd)) + 
  geom_bar(colour="black", alpha=0.3)

# Histology
# Does not make sense to pool nst and tubular? 
Histology[1:5]
unique(as.vector(Histology))
ggplot(pheno.df) +
  aes(Histology, fill=Histology) + 
  geom_bar(colour="black", alpha=0.3)
table(Histology)

# Histology1 - swop between lobular and other 
Histology1[1:5]
unique(as.vector(Histology1))
ggplot(pheno.df) +
  aes(Histology1, fill=Histology1) + 
  geom_bar(colour="black", alpha=0.3)
table(Histology1)

sum(as.vector(Histology) != as.vector(Histology1))

pheno.df %>% 
  select(labid.x, Histology, Histology1) %>% 
  filter(as.vector(Histology) != as.vector(Histology1))

#pheno.df[
#  as.vector(Histology) != as.vector(Histology1), 
#  c("labid.x","Histology","Histology1")]

# histol_cat
histo1_cat[1:5]
unique(as.vector(histo1_cat))

# Cases and controls are balanced by histology
ggplot(pheno.df) +
  aes(histo1_cat, fill=as.factor(cc.x)) + 
  geom_bar(colour="black", alpha=0.3)
table(cc.x, histo1_cat)
fisher.test(table(cc.x, histo1_cat))

# Balance within ductal and lobular: 
# non-significant trend to have more lobular in CBC
pheno.df %>% 
  filter(histo1_cat %in% c("ductal", "lobular")) %>% 
  ggplot() + 
    aes(histo1_cat, fill=as.factor(cc.x)) + 
    geom_bar(colour="black", alpha=0.3)

x <- table(
  as.vector(cc.x[histo1_cat == "ductal" | histo1_cat == "lobular"]), 
  as.vector(histo1_cat[histo1_cat == "ductal" | histo1_cat == "lobular"]))
x
fisher.test(x)
rm(x)

# Lobular only
lobular[1:5]
unique(as.vector(lobular))
ggplot(pheno.df) +
  aes(as.factor(lobular), fill=as.factor(cc.x)) +
  geom_bar(colour="black", alpha=0.3)
table(lobular)

# Hist_lob_other
Hist_lob_other[1:5]
unique(as.vector(Hist_lob_other))
ggplot(pheno.df) +
  aes(Hist_lob_other, fill=as.factor(cc.x)) +
  geom_bar(colour="black", alpha=0.3)
table(Hist_lob_other)

```

## receptors

check consistency with treatment?
much more more er+ ve?
many er unknown

```{r receptors}

# er1
er1[1:5]
class(er1)
unique(as.vector(er1))
ggplot(pheno.df) +
  aes(as.factor(er1), fill=as.factor(cc.x)) +
  geom_bar(colour="black", alpha=0.3)
summary(as.factor(er1))

# Balanced between cases and controls
table(cc.x, er1)
fisher.test(table(cc.x,er1))

ggplot(pheno.df) +
  aes(as.factor(cc.x), fill=as.factor(er1)) +
  geom_bar(colour="black", alpha=0.3)

# er1_num (=er1)
er1_num[1:5]
class(er1_num)
unique(as.vector(er1_num))
ggplot(pheno.df) +
  aes(as.factor(er1_num), fill=as.factor(cc.x)) +
  geom_bar(colour="black", alpha=0.3)

table(as.factor(er1_num))
sum(is.na(er1_num))
ggplot(pheno.df) +
  aes(as.factor(cc.x), fill=as.factor(er1_num)) +
  geom_bar(colour="black", alpha=0.3)

sum(as.vector(er1) != as.vector(er1), na.rm = TRUE)

# er1_cat
er1_cat[1:5]
class(er1_cat)
unique(as.vector(er1_cat))
ggplot(pheno.df) +
  aes(as.factor(er1_cat), fill=as.factor(cc.x)) +
  geom_bar(colour="black", alpha=0.3)
table(as.vector(er1_cat))
sum(is.na(er1_cat))

# Balanced between cases and controls
x <- table(cc.x, er1_cat)[,c(1,3,5)]
x
fisher.test(x)
rm(x)

# er_fd
er_fd[1:5]
class(er_fd)
unique(as.vector(er_fd))
ggplot(pheno.df) +
  aes(as.factor(er_fd), fill=as.factor(cc.x)) +
  geom_bar(colour="black", alpha=0.3)
table(as.vector(er_fd))
sum(is.na(er_fd))

# pr1_cat
pr1_cat[1:5]
class(pr1_cat)
unique(as.vector(pr1_cat))
ggplot(pheno.df) +
  aes(as.factor(pr1_cat), fill=as.factor(cc.x)) +
  geom_bar(colour="black", alpha=0.3)
table(as.vector(pr1_cat))
sum(is.na(pr1_cat))

# Balanced between cases and controls
x <- table(cc.x, pr1_cat)[,c(1,3,7)]
fisher.test(x)
rm(x)

ggplot(pheno.df) +
  aes(as.factor(cc.x), fill=as.factor(pr1_cat)) +
  geom_bar(colour="black", alpha=0.3)

# pr correlate with er
table(er1, pr1_cat)[,c(1,3)]
fisher.test(table(er1, pr1_cat)[,c(1,3)])

table(er1, pr1_cat)[,c(1,3,7)]
fisher.test(table(er1, pr1_cat)[,c(1,3,7)])

# Combined er and pr field
pheno.df <- pheno.df %>% mutate(epr=paste(er1_cat, pr1_cat))

# Look at combined faceted plots
ggplot(pheno.df) +
  aes( as.factor(cc.x), fill=as.factor(epr)) +
  geom_bar(colour="black", alpha=0.3)

ggplot(pheno.df) +
  aes( as.factor(cc.x), fill=as.factor(epr)) +
  geom_bar(colour="black", alpha=0.3) + 
  facet_grid(er1~pr1_cat)

# Excess of controls in ER-ve, PgR-ve
# Excess of controls in ER-ve, PgR+ve??

# Balanced cases and controls in ER+ve, PgR-ve
# Balanced cases and controls in ER+ve, PgR+ve

# ===================================================== #

# pr_fd
pr_fd[1:5]
class(pr_fd)
unique(as.vector(pr_fd))
ggplot(pheno.df) +
  aes(as.factor(pr_fd), fill=as.factor(cc.x)) +
  geom_bar(colour="black", alpha=0.3)
table(as.vector(pr_fd))
fisher.test(table(cc.x,pr_fd))
sum(is.na(pr_fd))

```

## reproductive_factors

```{r reproductive_factors}

# age_menarche (categorical)
age_menarche[1:5]
class(age_menarche)
unique(age_menarche)
ggplot(pheno.df) +
  aes(as.factor(age_menarche), fill=as.factor(cc.x)) +
  geom_bar(colour="black", alpha=0.3)
table(age_menarche)
sum(is.na(age_menarche))

# Balanced between cases and controls
table(cc.x, age_menarche)
fisher.test(table(cc.x,age_menarche))
sum(is.na(age_menarche))

ggplot(pheno.df) +
  aes(as.factor(cc.x), fill=as.factor(age_menarche)) +
  geom_bar(colour="black", alpha=0.3)

# Age menarche (numeric)
rh_age_menarche[1:5]
class(rh_age_menarche)
unique(rh_age_menarche)

hist(rh_age_menarche)
ggplot(pheno.df) + 
  aes(rh_age_menarche, fill=as.factor(cc.x)) +
  geom_histogram(colour="black", alpha=0.3, binwidth=1) +
  stat_bin(binwidth=1, geom="text", aes(label=..count..), vjust=-1)

ggplot(pheno.df) + 
  aes(rh_age_menarche, fill=as.factor(cc.x)) +
  geom_histogram(colour="black", alpha=0.3, binwidth=1)

table(cc.x, rh_age_menarche)
sum(is.na(rh_age_menarche))

# Balanced between cases and controls
t.test(rh_age_menarche~cc.x)

# odd density plot behaving ...
ggplot(pheno.df) + 
  aes(rh_age_menarche, fill=as.factor(cc.x)) +
  geom_density(colour="black", alpha=0.3)

# adjusted plot
ggplot(pheno.df) + 
  aes(rh_age_menarche, fill=as.factor(cc.x)) +
  geom_density(colour="black", alpha=0.3, adjust=3)

# age_menopause: 0,1,2 (pre, in, post?)

age_menopause[1:5]
class(age_menopause)
unique(age_menopause)

ggplot(pheno.df) +
  aes(as.factor(age_menopause), fill=as.factor(cc.x)) +
  geom_bar(colour="black", alpha=0.3)

table(age_menopause)
sum(is.na(age_menopause))

# Balanced cases and controls
table(cc.x, age_menopause)
fisher.test(table(cc.x, age_menopause))

ggplot(pheno.df) +
  aes(as.factor(cc.x), fill=as.factor(age_menopause)) +
  geom_bar(colour="black", alpha=0.3)

# age_menopause_1yrbf_fd: what is this? 
# Does not look as menopausal age...

age_menopause_1yrbf_fd[1:5]
class(age_menopause_1yrbf_fd)
unique(age_menopause_1yrbf_fd)

ggplot(pheno.df) +
  aes(age_menopause_1yrbf_fd, fill=as.factor(cc.x)) +
  geom_histogram(colour="black", alpha=0.3, binwidth=1)

table(age_menopause_1yrbf_fd)
sum(is.na(age_menopause_1yrbf_fd))

# Balanced between cases and controls
t.test(age_menopause_1yrbf_fd~cc.x)
pheno.df %>% 
  filter(age_menopause_1yrbf_fd != -1) %>% 
  t.test(data=., age_menopause_1yrbf_fd~cc.x)

pheno.df %>% 
  filter(age_menopause_1yrbf_fd != -1) %>% 
  ggplot() +
    aes(age_menopause_1yrbf_fd, fill=as.factor(cc.x)) +
    geom_histogram(colour="black", alpha=0.3, binwidth=1)

pheno.df %>% 
  filter(age_menopause_1yrbf_fd != -1) %>% 
  ggplot() +
    aes(age_menopause_1yrbf_fd, fill=as.factor(cc.x)) +
    geom_density(colour="black", alpha=0.3)

# Of those who is in mp there is trent to have younger mp age in controls
# This may be explained by more cytotoxic treatment in this group

# age_1fftp_fd: what is this? another age of a childbearing?
# ftp - in my prev life meant "full term pregnancy"?
# Less pregnancies in cases than in controls?

age_1fftp_fd[1:5]
class(age_1fftp_fd)
unique(age_1fftp_fd)

ggplot(pheno.df) + 
  aes(age_1fftp_fd, fill=as.factor(cc.x)) +
  geom_histogram(colour="black", alpha=0.3, binwidth=1)

ggplot(pheno.df) + 
  aes(age_1fftp_fd, fill=as.factor(cc.x)) +
  geom_density(colour="black", alpha=0.3)

pheno.df %>% filter(age_1fftp_fd!=0) %>% 
  ggplot() + 
  aes(age_1fftp_fd, fill=as.factor(cc.x)) +
  geom_density(colour="black", alpha=0.3)

pheno.df %>% filter(age_1fftp_fd!=0) %>% 
  t.test(data=.,age_1fftp_fd~cc.x)

table(age_1fftp_fd)
sum(is.na(age_1fftp_fd))

# Balanced between cases and controls
t.test(age_1fftp_fd~cc.x)
pheno.df %>% 
  filter(age_1fftp_fd != 0) %>% 
  t.test(data=., age_1fftp_fd~cc.x)

# Num_ftp_fd - what would it be?
# Num of full time pregnancies???
# The numnbers do not make sence and do not match the num_preg

Num_ftp_fd[1:5]
class(Num_ftp_fd)
unique(Num_ftp_fd)

ggplot(pheno.df) +
  aes(as.factor(Num_ftp_fd), fill=as.factor(cc.x)) +
  geom_bar(colour="black", alpha=0.3)

ggplot(pheno.df) +
  aes(as.factor(cc.x), fill=as.factor(Num_ftp_fd)) +
  geom_bar(colour="black", alpha=0.3)

table(cc.x, Num_ftp_fd)
sum(is.na(Num_ftp_fd))

# Balanced in cases and controls
#fisher.test(table(cc.x, Num_ftp_fd)) # can not calculate ...
fisher.test(table(cc.x, Num_ftp_fd)[,1:5]) # non-sign trend

# num_preg

num_preg[1:5]
class(num_preg)
unique(num_preg)

ggplot(pheno.df) +
  aes(as.factor(num_preg), fill=as.factor(cc.x)) +
  geom_bar(colour="black", alpha=0.3)

sum(is.na(num_preg))

# Less pregnancies in cases
table(cc.x, num_preg)
sum(is.na(num_preg))
fisher.test(table(cc.x, num_preg))

ggplot(pheno.df) +
  aes(as.factor(cc.x), fill=as.factor(num_preg)) +
  geom_bar(colour="black", alpha=0.3)

# No CBC with 3-pregnancies ...

```

## BMI
this is about lafelong exposure to Eg

in pre-menopause it is protective
this is about endocrine disorders and disruption of ms-cyclingand about cumulative 
lafelong exposure to Eg 
too speculative ...

```{r BMI}
# BMI_age18

BMI_age18[1:5]
class(BMI_age18)

ggplot(pheno.df) + 
  aes(BMI_age18, fill=as.factor(cc.x)) +
  geom_histogram(colour="black", alpha=0.3, binwidth=1)

ggplot(pheno.df) + 
  aes(BMI_age18, fill=as.factor(cc.x)) +
  geom_density(colour="black", alpha=0.3)

# controls tended to have slightly higher BMI at age of 18
# Caused by obese cases, consistent with protective role of
# obesity in pre-menopause
t.test(BMI_age18~cc.x)

summary(BMI_age18)
sum(is.na(BMI_age18))

# BMI_dx

BMI_dx[1:5]
class(BMI_dx)

ggplot(pheno.df) + 
  aes(BMI_dx, fill=as.factor(cc.x)) +
  geom_histogram(colour="black", alpha=0.3, binwidth=1)

ggplot(pheno.df) + 
  aes(BMI_dx, fill=as.factor(cc.x)) +
  geom_density(colour="black", alpha=0.3)

# controls had higher BMI at Ds
t.test(BMI_dx~cc.x)

summary(BMI_dx)
sum(is.na(BMI_dx))

# BMI_age18 vs BMI_dx

ggplot(pheno.df) + 
  aes(x=BMI_age18, y=BMI_dx, colour=as.factor(cc.x)) +
  geom_point(alpha=0.3) + 
  geom_smooth(method="lm")

# BMI_ref

BMI_ref[1:5]
class(BMI_ref)

ggplot(pheno.df) + 
  aes(BMI_ref, fill=as.factor(cc.x)) +
  geom_histogram(colour="black", alpha=0.3, binwidth=1)

ggplot(pheno.df) + 
  aes(BMI_ref, fill=as.factor(cc.x)) +
  geom_density(colour="black", alpha=0.3)

summary(BMI_ref)
sum(is.na(BMI_ref))

# BMI_ref vs BMI_dx
ggplot(pheno.df) + 
  aes(x=BMI_ref, y=BMI_dx, colour=as.factor(cc.x)) +
  geom_point(alpha=0.3) + 
  geom_smooth(method="lm")

```

## general_data_fields

```{r general_data_fields}

# DOB

DOB[1:5]
class(DOB)
hist(DOB)

ggplot(pheno.df) + 
  aes(DOB, fill=as.factor(cc.x)) +
  geom_histogram(colour="black", alpha=0.3)

ggplot(pheno.df) + 
  aes(DOB, fill=as.factor(cc.x)) +
  geom_density(colour="black", alpha=0.3)

# dxyr_stratum

dxyr_stratum[1:5]
class(dxyr_stratum)
summary(dxyr_stratum)

ggplot(pheno.df) + 
  aes(dxyr_stratum, fill=as.factor(cc.x)) +
  geom_bar(colour="black", alpha=0.3)

# Absolutely balanced between cases and controls
table(dxyr_stratum)
sum(is.na(dxyr_stratum))
fisher.test(table(cc.x,dxyr_stratum))

ggplot(pheno.df) + 
  aes(as.factor(cc.x), fill=as.factor(dxyr_stratum)) +
  geom_bar(colour="black", alpha=0.3)

# age_stratum1

age_stratum1[1:5]
class(age_stratum1)
summary(age_stratum1)

ggplot(pheno.df) + 
  aes(age_stratum1, fill=as.factor(cc.x)) +
  geom_bar(colour="black", alpha=0.3)

table(age_stratum1)
sum(is.na(age_stratum1))

# Absolutely balanced between cases and controls
table(age_stratum1)
sum(is.na(age_stratum1))
fisher.test(table(cc.x,age_stratum1))

ggplot(pheno.df) + 
  aes(as.factor(cc.x), fill=as.factor(age_stratum1)) +
  geom_bar(colour="black", alpha=0.3)

# sub_dx_age_cat

sub_dx_age_cat[1:5]
class(sub_dx_age_cat)
summary(sub_dx_age_cat)

ggplot(pheno.df) + 
  aes(sub_dx_age_cat, fill=as.factor(cc.x)) +
  geom_bar(colour="black", alpha=0.3)

table(sub_dx_age_cat)
sum(is.na(sub_dx_age_cat))

# Balanced between cases and controls
fisher.test(table(cc.x,sub_dx_age_cat))

ggplot(pheno.df) + 
  aes(as.factor(cc.x), fill=as.factor(sub_dx_age_cat)) +
  geom_bar(colour="black", alpha=0.3)

# refage

refage[1:5]
class(refage)
summary(refage)

ggplot(pheno.df) + 
  aes(refage, fill=as.factor(cc.x)) +
  geom_histogram(colour="black", alpha=0.3, binwidth=1)

ggplot(pheno.df) + 
  aes(refage, fill=as.factor(cc.x)) +
  geom_density(colour="black", alpha=0.3)

summary(refage)
sum(is.na(refage))

# balanced in cases and controls
t.test(refage~cc.x)

# rstime

rstime[1:5]
class(rstime)
summary(rstime)

ggplot(pheno.df) + 
  aes(rstime, fill=as.factor(cc.x)) +
  geom_histogram(colour="black", alpha=0.3, binwidth=1)

ggplot(pheno.df) + 
  aes(rstime, fill=as.factor(cc.x)) +
  geom_density(colour="black", alpha=0.3)
sum(is.na(rstime))

# balanced in cases and controls
t.test(refage~cc.x)

```

## explore_rstime_and_offset

```{r explore_rstime_and_offset}

ggplot(pheno.df) + 
  aes(offset.x, fill=as.factor(cc.x)) +
  geom_histogram(colour="black", alpha=0.3, binwidth=1)

ggplot(pheno.df) + 
  geom_density(aes(offset.x), fill="red", colour="black", alpha=0.3) +
  geom_density(aes(rstime), fill="blue", colour="black", alpha=0.3)

ggplot(pheno.df) + 
  geom_density(aes(rstime), fill="red", colour="black", alpha=0.3) +
  geom_density(aes(refage - sub_dx_age.x), fill="blue", colour="black", alpha=0.3)

ggplot(pheno.df) + 
  aes(refage - sub_dx_age.x, fill=as.factor(cc.x)) +
  geom_histogram(colour="black", alpha=0.3, binwidth=1)

ggplot(pheno.df) + 
  aes(refage - sub_dx_age.x, fill=as.factor(cc.x)) +
  geom_density(colour="black", alpha=0.3)

ggplot(pheno.df) + 
  geom_density(aes(refage), fill="red", colour="black", alpha=0.3) +
  geom_density(aes(sub_dx_age.x), fill="blue", colour="black", alpha=0.3)

```

## unknown_fields

```{r unknown_fields}

# dsmiss

dsmiss[1:5]
class(dsmiss)
summary(dsmiss)

table(dsmiss)
sum(is.na(dsmiss))

ggplot(pheno.df) + 
  aes(dsmiss, fill=as.factor(cc.x)) +
  geom_bar(colour="black", alpha=0.3)

# Balanced between cases and controls
fisher.test(table(cc.x,dsmiss))

ggplot(pheno.df) + 
  aes(as.factor(cc.x), fill=as.factor(dsmiss)) +
  geom_bar(colour="black", alpha=0.3)

# conf_miss

conf_miss[1:5]
class(conf_miss)
summary(conf_miss)

table(conf_miss)
sum(is.na(conf_miss))

ggplot(pheno.df) + 
  aes(conf_miss, fill=as.factor(cc.x)) +
  geom_bar(colour="black", alpha=0.3)

# Balanced between cases and controls
fisher.test(table(cc.x,conf_miss))

ggplot(pheno.df) + 
  aes(as.factor(cc.x), fill=as.factor(conf_miss)) +
  geom_bar(colour="black", alpha=0.3)

# dsmiss vs conf_miss

sum(dsmiss != conf_miss, na.rm=TRUE)

# Phase - all 1

Phase[1:5]
class(Phase)
summary(Phase)

table(Phase)
sum(is.na(Phase))
sum(Phase!=1)

# Deleterious

Deleterious[1:5]
class(Deleterious)
summary(Deleterious)

table(Deleterious)
sum(is.na(Deleterious))

ggplot(pheno.df) + 
  aes(as.factor(Deleterious), fill=as.factor(cc.x)) +
  geom_bar(colour="black", alpha=0.3)

# Balanced between cases and controls
fisher.test(table(cc.x,Deleterious))

ggplot(pheno.df) + 
  aes(as.factor(cc.x), fill=as.factor(Deleterious)) +
  geom_bar(colour="black", alpha=0.3)

# Detach pheno data frame
detach(pheno.df)

```

# reshape_pheno_table

```{r reshape_pheno_table}

# Select columns
pheno.df <- pheno.df %>% select(
  labid = labid.x,
  cc = cc.x,
  family_history = family_history.y,
  sub_dx_age = sub_dx_age.x,
  refage = refage,
  rstime = rstime,
  offset = offset.x,
  Eigen_1 = Eigen_1.x,
  Eigen_2 = Eigen_2.x,
  Eigen_3 = Eigen_3.x,
  Eigen_4 = Eigen_4.x,
  Eigen_5 = Eigen_5.x,
  stage_fd = stage_fd,
  er1 = er1,
  pr1_cat = pr1_cat,
  histo1_cat = histo1_cat,
  hormone = hormone,
  CMF = CMF,
  XRTBrCHAR = XRTBrCHAR,
  dose_caseloc = dose_caseloc,
  rh_age_menarche = rh_age_menarche,
  age_1fftp_fd = age_1fftp_fd, 
  age_menopause_1yrbf_fd = age_menopause_1yrbf_fd,
  num_preg = num_preg,
  BMI_age18 = BMI_age18,
  BMI_dx = BMI_dx,
  BMI_ref = BMI_ref)

# Recode categorical pr to NA, 0 and 1
pheno.df <- pheno.df %>% mutate(
  pr1 = ifelse(pr1_cat=="negative", 0,
        ifelse(pr1_cat=="positive", 1,
        NA)))

# Substitute zero age of menarche to NA
pheno.df <- pheno.df %>% mutate(
  age_menarche = ifelse(rh_age_menarche==0, NA, rh_age_menarche))

# Remove corrected and recoded fields
pheno.df <- pheno.df %>% select(-rh_age_menarche, -pr1_cat)

# Remove demographics table
rm(demographics.df, odd_cases, duplicated_cases)

```

# merge_pheno_and_covar

```{r merge_pheno_and_covar}

# Check data frames
dim(covar.df)
dim(pheno.df)
colnames(covar.df)
colnames(pheno.df)

# Prepare key for joining
pheno.df$labid <- as.vector(pheno.df$labid)
covar.df$labid <- as.vector(covar.df$labid)

ph.df <- left_join(covar.df, pheno.df, by="labid")
dim(ph.df)
colnames(ph.df)

attach(ph.df)

sum(cc.x != cc.y, na.rm = TRUE)
sum(is.na(cc.x)) # to be preserved
sum(is.na(cc.y)) # to be removed

sum(sub_dx_age.x != sub_dx_age.y, na.rm = TRUE)
sum(is.na(sub_dx_age.x)) # to be preserved
sum(is.na(sub_dx_age.y)) # to be removed

sum(offset.x != offset.y, na.rm = TRUE)
sum(is.na(offset.x)) # to be preserved
sum(is.na(offset.y)) # to be removed

sum(hormone.x != hormone.y, na.rm = TRUE)
sum(is.na(hormone.x)) # to be preserved
sum(is.na(hormone.y)) # to be removed

sum(is.na(chemo)) # to be removed, substituted by CMF

sum(Eigen_1.x != Eigen_1.y, na.rm = TRUE)
sum(is.na(Eigen_1.x)) # to be preserved
sum(is.na(Eigen_1.y)) # to be removed

sum(Eigen_2.x != Eigen_2.y, na.rm = TRUE)
sum(is.na(Eigen_2.x)) # to be preserved
sum(is.na(Eigen_2.y)) # to be removed

sum(Eigen_3.x != Eigen_3.y, na.rm = TRUE)
sum(is.na(Eigen_3.x)) # to be preserved
sum(is.na(Eigen_3.y)) # to be removed

sum(Eigen_4.x != Eigen_4.y, na.rm = TRUE)
sum(is.na(Eigen_4.x)) # to be preserved
sum(is.na(Eigen_4.y)) # to be removed

sum(Eigen_5.x != Eigen_5.y, na.rm = TRUE)
sum(is.na(Eigen_5.x)) # to be preserved
sum(is.na(Eigen_5.y)) # to be removed

sum(is.na(family_history)) # to be preserved
sum(is.na(refage)) # to be preserved
sum(is.na(rstime)) # to be preserved

sum(is.na(stage_fd)) # to be preserved
sum(is.na(er1)) # to be preserved
sum(is.na(histo1_cat)) # to be preserved
sum(is.na(CMF)) # to be preserved
sum(is.na(XRTBrCHAR)) # to be preserved
sum(is.na(dose_caseloc)) # to be preserved
sum(is.na(age_1fftp_fd)) # to be preserved
sum(is.na(age_menopause_1yrbf_fd)) # to be preserved
sum(is.na(num_preg)) # to be preserved
sum(is.na(BMI_age18)) # to be preserved
sum(is.na(BMI_dx)) # to be preserved
sum(is.na(BMI_ref)) # to be preserved
sum(is.na(pr1)) # to be preserved
sum(is.na(age_menarche)) # to be preserved

detach(ph.df)

phen_types.df <- ph.df %>% 
  select(
    gwas_id = labid, 
    cc = cc.x, 
    family_history = family_history,
    age_dx = sub_dx_age.x, 
    age_ref = refage, 
    rstime = rstime,
    offset = offset.x, 
    Eigen_1 = Eigen_1.x, 
    Eigen_2 = Eigen_2.x, 
    Eigen_3 = Eigen_3.x, 
    Eigen_4 = Eigen_4.x, 
    Eigen_5 = Eigen_5.x, 
    stage = stage_fd, 
    er1 = er1, 
    pr1 = pr1, 
    hist_cat = histo1_cat, 
    hormone = hormone.x, 
    chemo_cat = CMF, 
    br_xray = XRTBrCHAR, 
    br_xray_dose = dose_caseloc, 
    age_menarche = age_menarche, 
    age_1st_ftp = age_1fftp_fd, 
    age_menopause = age_menopause_1yrbf_fd, 
    num_preg = num_preg, 
    BMI_age18 = BMI_age18, 
    BMI_dx = BMI_dx, 
    BMI_ref = BMI_ref)

rm(ph.df, pheno.df, covar.df)

```

# merge_phenotypes_and_samples

```{r merge_phenotypes_and_samples}

phenotypes.df <- full_join(samples.df, phen_types.df, by="gwas_id")

dim(phenotypes.df)
colnames(phenotypes.df)
str(phenotypes.df)
phenotypes.df[1:10, 1:7]

rm(samples.df, phen_types.df)

```

# data_summary

```{r data_summary}

ls()

dim(genotypes.mx)
genotypes.mx[1:5,1:5]

dim(phenotypes.df)
colnames(phenotypes.df)
phenotypes.df[1:5,1:5]

dim(variants.df)
colnames(variants.df)
variants.df[1:5,1:5]

# check consistency of samples in genotypes and phenotypes
genotype_samples <- colnames(genotypes.mx)
length(genotype_samples)

all_phenotype_samples <- 
  phenotypes.df %>% select(wes_id)
all_phenotype_samples <- as.vector(all_phenotype_samples[,1])
length(all_phenotype_samples)

pass_phenotype_samples<- 
  phenotypes.df %>% filter(filter=="pass") %>% select(wes_id)
pass_phenotype_samples <- as.vector(pass_phenotype_samples[,1])
length(pass_phenotype_samples)

sum( genotype_samples %in% all_phenotype_samples )
sum( genotype_samples %in% pass_phenotype_samples )

rm(genotype_samples, all_phenotype_samples, pass_phenotype_samples)

# check consistency of variants in genotypes and variants
sum(rownames(genotypes.mx) != variants.df$VarID )

```

# save_data

```{r save_data}

save.image(file=interim_data_file)

```

# final_section

```{r final_section}
ls()
sessionInfo()
Sys.time()
```
